core/src/main/scala/org/bitcoins/core/api/Callback.scala

```diff
@@ -0,0 +1,56 @@
+package org.bitcoins.core.api
+
+import org.bitcoins.core.util.{FutureUtil, SeqWrapper}
+import org.slf4j.Logger
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
+/** A function to be called in response to an event */
+trait Callback[T] {
+  def apply(param: T): Future[Unit]
+}
+
+/** A function with two parameters to be called in response to an event */
+trait CallbackBinary[T1, T2] extends Callback[(T1, T2)] {
+  def apply(param1: T1, param2: T2): Future[Unit]
+
+  override def apply(param: (T1, T2)): Future[Unit] = apply(param._1, param._2)
+}
+
+/** A function with three parameters to be called in response to an event */
+trait CallbackTernary[T1, T2, T3] extends Callback[(T1, T2, T3)] {
+  def apply(param1: T1, param2: T2, param3: T3): Future[Unit]
+
+  override def apply(param: (T1, T2, T3)): Future[Unit] =
+    apply(param._1, param._2, param._3)
+}
+
+/** Manages a set of callbacks, should be used to manage execution and logging if needed */
+case class CallbackHandler[C, T <: Callback[C]](
+    name: String,
+    override val wrapped: IndexedSeq[T])
+    extends SeqWrapper[T] {
+
+  def ++(other: CallbackHandler[C, T]): CallbackHandler[C, T] = {
+    require(name == other.name,
+            "Cannot combine callback handlers with different names")
+    CallbackHandler(name, wrapped ++ other.wrapped)
+  }
+
+  def execute(param: C, recoverFunc: Throwable => Unit = _ => ())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+    FutureUtil.foldLeftAsync((), wrapped)((_, callback) =>
```

Comment:
There is still a problem here with using `foldLeftAsync()`. I'd be curious to here @rorp 's opinion on this:

Let's say we have two callbacks defined as this: 

```scala
  val f1 = new Callback[Int] {
    def apply(param: Int): Future[Unit] = {
      Thread.sleep(1000000) //sleep 1,000 seconds for whatever reason (simulated delay)
      Future.unit
    }
  }
  
  val f2 = new Callback[Int] {
    def apply(param: Int): Future[Unit] = {
      //no delay here, execute right away
      Future.unit
    }
  }
```

We will now wait for 1,000 seconds on `f1` before we start executing `f2`. Is this really the behavior we want?

Reply:
I see your point now, maybe have a CallbackHandler and a separate SynchronousCallbackHandler

Reply:
I'm not really understand why use `Future`s this way. If we want to wait for 1000 seconds we need to put `sleep()` inside a `Future`: 

```scala
  val f1 = new Callback[Int] {
    def apply(param: Int): Future[Unit] = Future {
      Thread.sleep(1000000) //sleep 1,000 seconds for whatever reason (simulated delay)
    }
  }
```

Reply:
I think we should just say there are no execution guarantees when passing in the callbacks. You can bake in ordering if you want, by doing something like this: 

```scala
  val f3 = new Callback[Int] {
    def apply(param: Int): Future[Unit] = {
      //since i care about ordering, i need to enforce ordering in this callback
      //since it's important to call f1 before f2
      
      f1(param).flatMap(_ =>  f2(param))
    }
  }
```

Reply:
> I think we should just say there are no execution guarantees when passing in the callbacks. 
I still don't understand what's the issue here. `foldLeftAsync()` in fact has execution guarantees -- it executes `Future`s one by one. If one doesn't want the Future`s to be executed one by one should simply not wrap the code in `Future`s or not return them outside `Callback.apply()`.

I think that's what we should say.